filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsResolver.scala

```diff
@@ -1,28 +1,36 @@
 package cromwell.filesystems.drs
 
-import cloud.nio.impl.drs.{DrsCloudNioFileSystemProvider, MarthaResponse, Url}
+import cats.effect.IO
+import cloud.nio.impl.drs.{DrsCloudNioFileProvider, DrsCloudNioFileSystemProvider, Url}
+import org.apache.commons.lang3.exception.ExceptionUtils
+import shapeless.syntax.typeable._
 
 
 object DrsResolver {
   private val GcsScheme: String = "gs"
 
-  private def extractPathRelativeToScheme(drsPath: String, urlArray: Array[Url], scheme: String): String = {
-    val schemeUrlOption = urlArray.find(u => u.url.startsWith(scheme))
+
+  def extractUrlForScheme(urlArray: Array[Url], scheme: String): Either[UrlNotFoundException, String] = {
+    val schemeUrlOption = urlArray.find(_.url.startsWith(scheme))
 
     schemeUrlOption match {
-      case Some(schemeUrl) => schemeUrl.url.substring(scheme.length + 3)
-      case None => throw GcsUrlNotFoundException(drsPath, scheme)
+      case Some(schemeUrl) => Right(schemeUrl.url)
+      case None => Left(UrlNotFoundException(scheme))
     }
   }
 
-  def getContainerRelativePath(drsPath: DrsPath): String = {
-    val drsFileSystemProvider = drsPath.drsPath.getFileSystem.provider.asInstanceOf[DrsCloudNioFileSystemProvider]
-    val marthaResponseObj: MarthaResponse = drsFileSystemProvider.drsPathResolver.resolveDrsThroughMartha(drsPath.pathAsString)
+  def getContainerRelativePath(drsPath: DrsPath): IO[String] = {
+    val drsFileSystemProviderOption = drsPath.drsPath.getFileSystem.provider.cast[DrsCloudNioFileSystemProvider]
+
+    val noFileSystemForDrsError = s"Unable to cast file system provider to DrsCloudNioFileSystemProvider for DRS path $drsPath."
 
-    //Currently, Martha only supports resolving DRS paths to GCS paths
-    extractPathRelativeToScheme(drsPath.pathAsString, marthaResponseObj.dos.data_object.urls, GcsScheme)
+    for {
+      drsFileSystemProvider <- DrsCloudNioFileProvider.convertOptionToIOWithErrorMsg(drsFileSystemProviderOption, noFileSystemForDrsError)
+      marthaResponse <- drsFileSystemProvider.drsPathResolver.resolveDrsThroughMartha(drsPath.pathAsString)
+      //Currently, Martha only supports resolving DRS paths to GCS paths
+      relativePath <- IO.fromEither(extractUrlForScheme(marthaResponse.dos.data_object.urls, GcsScheme))
+        .map(_.substring(GcsScheme.length + 3))
```

Comment:
maybe pull this out to a documented constant, its usage here is a bit magical

Reply:
I have extracted `scheme.length + 3` into a method. This is because the scheme is eventually going to be dynamic (it won't always be gcs) and it was hard to describe what 3 stands for, which is length of `://`.